File: legacy_python_app/src/transcription/whisper.py
import os
import threading
import time
from functools import wraps

import dotenv
import httpx
#from openai import OpenAI
from opencc import OpenCC

#from ..llm.symbol import SymbolProcessor
from ..utils.logger import logger

dotenv.load_dotenv()

def timeout_decorator(seconds):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            result = [None]
            error = [None]
            completed = threading.Event()

            def target():
                try:
                    result[0] = func(*args, **kwargs)
                except Exception as e:
                    error[0] = e
                finally:
                    completed.set()

            thread = threading.Thread(target=target)
            thread.daemon = True
            thread.start()

            if completed.wait(seconds):
                if error[0] is not None:
                    raise error[0]
                return result[0]
            raise TimeoutError(f"操作超时 ({seconds}秒)")

        return wrapper
    return decorator

class WhisperProcessor:
    # 类级别的配置参数
    DEFAULT_TIMEOUT = 20  # API 超时时间（秒）
    DEFAULT_MODEL = None
    
    def __init__(self):
        api_key = os.getenv("GROQ_API_KEY")
        base_url = os.getenv("GROQ_BASE_URL")
        self.convert_to_simplified = os.getenv("CONVERT_TO_SIMPLIFIED", "false").lower() == "true"
        self.cc = OpenCC('t2s') if self.convert_to_simplified else None
        #self.symbol = SymbolProcessor()
        self.add_symbol = os.getenv("ADD_SYMBOL", "false").lower() == "true"
        self.optimize_result = os.getenv("OPTIMIZE_RESULT", "false").lower() == "true"
        self.timeout_seconds = self.DEFAULT_TIMEOUT
        self.service_platform = os.getenv("SERVICE_PLATFORM", "groq").lower()

        if self.service_platform == "groq":
            raise ValueError(f"未知的平台: {self.service_platform}")
            assert api_key, "未设置 GROQ_API_KEY 环境变量"
            self.client = OpenAI(
                api_key=api_key,
                base_url=base_url if base_url else None
            )
            self.DEFAULT_MODEL = "whisper-large-v3-turbo"
        elif self.service_platform == "siliconflow":
            assert api_key, "未设置 SILICONFLOW_API_KEY 环境变量"
            self.DEFAULT_MODEL = "FunAudioLLM/SenseVoiceSmall"
        else:
            raise ValueError(f"未知的平台: {self.service_platform}")

    def _convert_traditional_to_simplified(self, text):
        """将繁体中文转换为简体中文"""
        if not self.convert_to_simplified or not text:
            return text
        return self.cc.convert(text)
    
    @timeout_decorator(10)
    def _call_whisper_api(self, mode, audio_data, prompt):
        """调用 Whisper API"""
        if mode == "translations":
            response = self.client.audio.translations.create(
                model="whisper-large-v3",
                response_format="text",
                prompt=prompt,
                file=("audio.wav", audio_data)
            )
        else:  # transcriptions
            response = self.client.audio.transcriptions.create(
                model="whisper-large-v3-turbo",
                response_format="text",
                prompt=prompt,
                file=("audio.wav", audio_data)
            )
        return str(response).strip()

    def process_audio(self, audio_buffer, mode="transcriptions", prompt=""):
        """调用 Whisper API 处理音频（转录或翻译）
        
        Args:
            audio_path: 音频文件路径
            mode: 'transcriptions' 或 'translations'，决定是转录还是翻译
            prompt: 提示词
        
        Returns:
            tuple: (结果文本, 错误信息)
            - 如果成功，错误信息为 None
            - 如果失败，结果文本为 None
        """
        try:
            start_time = time.time()

            logger.info(f"正在调用 Whisper API... (模式: {mode})")
            result = self._call_whisper_api(mode, audio_buffer, prompt)

            logger.info(f"API 调用成功 ({mode}), 耗时: {time.time() - start_time:.1f}秒")
            result = self._convert_traditional_to_simplified(result)
            logger.info(f"识别结果: {result}")
            
            # 仅在 groq API 时添加标点符号
            if self.service_platform == "groq" and self.add_symbol:
                result = self.symbol.add_symbol(result)
                logger.info(f"添加标点符号: {result}")
            if self.optimize_result:
                result = self.symbol.optimize_result(result)
                logger.info(f"优化结果: {result}")

            return result, None
            

        except TimeoutError:
            error_msg = f"❌ API 请求超时 ({self.timeout_seconds}秒)"
            logger.error(error_msg)
            return None, error_msg
        except Exception as e:
            error_msg = f"❌ {str(e)}"
            logger.error(f"音频处理错误: {str(e)}", exc_info=True)
            return None, error_msg
        finally:
            audio_buffer.close()  # 显式关闭字节流

File: src-tauri/src/voice_assistant/asr/whisper.rs
use std::io::Cursor;
use reqwest::multipart;
use crate::voice_assistant::{AsrProcessor, Mode, VoiceError};
use serde_json::Value;
use std::time::Duration;

pub struct WhisperProcessor {
    client: reqwest::Client,
    api_key: String,
    base_url: String,
    convert_to_simplified: bool,
    add_symbol: bool,
    optimize_result: bool,
}

impl WhisperProcessor {
    pub fn new() -> Result<Self, VoiceError> {
        let api_key = std::env::var("GROQ_API_KEY")
            .map_err(|_| VoiceError::Other("GROQ_API_KEY environment variable not set".to_string()))?;
        
        let base_url = std::env::var("GROQ_BASE_URL")
            .unwrap_or_else(|_| "https://api.groq.com".to_string());

        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .map_err(|e| VoiceError::Network(e))?;

        Ok(Self {
            client,
            api_key,
            base_url,
            convert_to_simplified: std::env::var("CONVERT_TO_SIMPLIFIED")
                .unwrap_or_else(|_| "true".to_string())
                .parse()
                .unwrap_or(true),
            add_symbol: std::env::var("ADD_SYMBOL")
                .unwrap_or_else(|_| "true".to_string())
                .parse()
                .unwrap_or(true),
            optimize_result: std::env::var("OPTIMIZE_RESULT")
                .unwrap_or_else(|_| "true".to_string())
                .parse()
                .unwrap_or(true),
        })
    }

    async fn call_whisper_api(
        &self,
        mode: Mode,
        audio_data: &[u8],
        prompt: &str,
    ) -> Result<String, VoiceError> {
        let model = match mode {
            Mode::Transcriptions => "whisper-large-v3-turbo",
            Mode::Translations => "whisper-large-v3",
        };

        let form = multipart::Form::new()
            .part("file", multipart::Part::bytes(audio_data.to_vec())
                .file_name("audio.wav")
                .mime_str("audio/wav")?)
            .text("model", model)
            .text("response_format", "json");

        let form = if !prompt.is_empty() {
            form.text("prompt", prompt.to_string())
        } else {
            form
        };

        let request = self.client
            .post(&format!("{}/openai/v1/audio/{}", self.base_url, 
                if mode == Mode::Translations { "translations" } else { "transcriptions" }))
            .header("Authorization", format!("Bearer {}", self.api_key))
            .multipart(form);

        let response = request
            .send()
            .await
            .map_err(|e| VoiceError::Network(e))?;

        let status = response.status();
        if !status.is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(VoiceError::Other(format!("Whisper API error: {} - {}", status, error_text)));
        }

        let result: Value = response.json().await
            .map_err(|e| VoiceError::Network(e))?;

        if let Some(text) = result.get("text").and_then(|v| v.as_str()) {
            let mut processed_text = text.to_string();
            
            if self.add_symbol {
                processed_text = self.add_punctuation(&processed_text);
            }
            
            if self.optimize_result {
                processed_text = self.optimize_text(&processed_text);
            }
            
            if self.convert_to_simplified {
                processed_text = self.convert_traditional_to_simplified(&processed_text);
            }
            
            Ok(processed_text)
        } else {
            Err(VoiceError::Other("No text in Whisper response".to_string()))
        }
    }

    fn add_punctuation(&self, text: &str) -> String {
        // Simple punctuation addition - in a real implementation,
        // this could use more sophisticated NLP
        text.to_string()
    }

    fn optimize_text(&self, text: &str) -> String {
        // Basic text optimization - remove excessive whitespace
        text.split_whitespace().collect::<Vec<_>>().join(" ")
    }

    fn convert_traditional_to_simplified(&self, text: &str) -> String {
        // Placeholder for Chinese text conversion
        // In a real implementation, you'd use a library like chinese-conversion
        text.to_string()
    }
}

impl AsrProcessor for WhisperProcessor {
    fn process_audio(
        &self,
        audio_buffer: Cursor<Vec<u8>>,
        mode: Mode,
        prompt: &str,
    ) -> Result<String, VoiceError> {
        // Since we need async, but the trait is sync, we'll use a blocking runtime
        let rt = tokio::runtime::Runtime::new()
            .map_err(|e| VoiceError::Other(format!("Failed to create runtime: {}", e)))?;

        rt.block_on(async {
            let audio_data = audio_buffer.into_inner();
            self.call_whisper_api(mode, &audio_data, prompt).await
        })
    }
}